sysrib/sysrib.go

```diff
@@ -0,0 +1,408 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package sysrib implements a system-level RIB that is populated initially using
+// an OpenConfig configuration.
+package sysrib
+
+import (
+	"errors"
+	"fmt"
+	"net"
+	"sort"
+	"sync"
+
+	"github.com/kentik/patricia"
+	"github.com/kentik/patricia/generics_tree"
+	"github.com/openconfig/gribigo/afthelper"
+	oc "github.com/openconfig/gribigo/ocrt"
+	"github.com/openconfig/ygot/ytypes"
+)
+
+// SysRIB is a RIB data structure that can be used to resolve routing entries to their egress interfaces.
+// Currently it supports only IPv4 entries.
+type SysRIB struct {
+	// mu protects the map of network instance RIBs.
+	mu sync.RWMutex
+	// NI is the list of network instances (aka VRFs)
+	NI        map[string]*NIRIB
+	defaultNI string
+}
+
+// NIRIB is the RIB for a single network instance.
+type NIRIB struct {
+	// IPv4 is the IPv4 RIB
+	IPv4 *generics_tree.TreeV4[*Route]
+}
+
+type RoutePreference struct {
+	// AdminDistance is the admin distance of the protocol that added this
+	// route.
+	AdminDistance uint8 `json:"admin-distance"`
+	// Metric is the metric of the route. It is comparable only within
+	// routes of the same protocol, and therefore the same admin distance.
+	Metric uint32 `json:"metric"`
+}
+
+// Route is used to store a route in the radix tree.
+type Route struct {
+	// Prefix is a prefix that was being stored.
+	Prefix string `json:"prefix"`
+	// Connected indicates that the route is directly connected.
+	Connected *Interface `json:"connected"`
+	// NextHops is the set of IP nexthops that the route uses if
+	// it is not a connected route.
+	NextHops  []*afthelper.NextHopSummary `json:"nexthops"`
+	RoutePref RoutePreference
+}
+
+// NewSysRIB returns a SysRIB from an input parsed OpenConfig configuration.
+func NewSysRIB(cfg *oc.Device) (*SysRIB, error) {
```

Comment:
We probably have to figure out - as a TODO - how the configuration handling is going to work. We'll have `gnmit` (settable) that will be storing config + telemetry data. It seems like in the end-state design we want is to have sysrib subscribe to there - such that when the config comes in we parse it - write it to gnmit's cache, and then the RIB can determine what the connected routes are based on those subscriptions.

No need to refactor at the moment, just thinking in text.

Reply:
This sounds good.